Allow omitting initializer for Final attributes in dataclasses

Dataclasses typically don't have `__init__`, so this is rejected, but is probably valid:
```python
@dataclass
class C:
    x: int
    y: Final[int]
```
A possible solution is that plugin should set the `final_set_in_init` flag manually for the corresponding field `Var`s.

I just ran into this. `pyright` allows these type annotations as is (`Final` indicates an instance variable), and discussion in https://github.com/python/cpython/issues/89547 seems to indicate this is what'd be most ergonomic to support. It also matches how `dataclasses` itself handles these annotations.

For now, a workaround is to add `# type: ignore` on the annotation line. The dataclass definition still seems to be processed correctly, this just skips the error about a missing initializer.

```python
import dataclasses
from typing import Final


@dataclasses.dataclass
class Test:
    # The `type: ignore` is needed here to avoid an error due to a missing
    # initializer
    x: Final[int]  # type: ignore


t = Test(1)  # Signature still inferred properly
reveal_type(t.x)
t2 = Test("bad")  # Which means mypy marks this as an error
```

```
$ mypy ex.py 
ex.py:13: note: Revealed type is "builtins.int"
ex.py:14: error: Argument 1 to "Test" has incompatible type "str"; expected "int"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

For my own use cases, it'd be nice if the fix here also applied to the recent `dataclass_tranform` support, so this fix could work for other dataclass-like things as well.

--- a/mypy/plugins/dataclasses.py
+++ b/mypy/plugins/dataclasses.py
@@ -660,6 +660,14 @@ class DataclassTransformer:
                 else:
                     self._api.fail('"kw_only" argument must be a boolean literal', stmt.rvalue)
 
+            # Allow bare x: Final[int] in class body, since it will be set in the generated
+            # __init__() method (unless it is an InitVar), to match regular class semantics.
+            if node.is_final and not node.is_classvar and node.final_unset_in_class:
+                if is_init_var:
+                    self._api.fail("InitVar cannot be final", stmt.rvalue)
+                else:
+                    node.final_set_in_init = True
+
             if sym.type is None and node.is_final and node.is_inferred:
                 # This is a special case, assignment like x: Final = 42 is classified
                 # annotated above, but mypy strips the `Final` turning it into x = 42.

--- a/test-data/unit/check-dataclasses.test
+++ b/test-data/unit/check-dataclasses.test
@@ -2698,3 +2698,30 @@ class ClassB(ClassA):
     def value(self) -> int:
         return 0
 [builtins fixtures/dict.pyi]
+
+[case testDataclassAllowsFinalField]
+# flags: --python-version=3.13
+from dataclasses import InitVar, dataclass
+from typing import Final, ClassVar
+
+@dataclass
+class C:
+    x: int
+    y: Final[int]
+    z: ClassVar[Final[int]] = 4
+
+c = C(1, 2)
+c.y = 3  # E: Cannot assign to final attribute "y"
+
+@dataclass
+class D(C):
+    x: Final[int]  # E: Cannot override writable attribute "x" with a final one
+
+@dataclass
+class Bad1:
+    x: Final[InitVar[int]]  # E: InitVar cannot be final
+
+@dataclass
+class Bad2:
+    x: InitVar[Final[int]]  # E: Final can be only used as an outermost qualifier in a variable annotation
+[builtins fixtures/dict.pyi]

